### ISSUE TITLE
Regression in `ares_getaddrinfo` on the A and AAAA query order.

### ISSUE THREAD

**Issue by ChristianAmmer** [c-ares/c-ares#285]
With #257 the order search domains for T_A and T_AAAA requests are tried has changed.

*Before:*
T_AAAA for host.suffix1 and parallel T_A for host.suffix1
T_AAAA for host.suffix2 and parallel T_A for host.suffix2

*After:*
T_AAAA for host.suffix1
T_AAAA for host.suffix2
T_A for host.suffix1
T_A for host.suffix2

This slows down the entire lookup, in my opinion a regression.
In comparison to this, glibc performs IPv4 and IPv6 lookups in parallel with the intention of not slowing down the process (see [man resolv.conf](http://man7.org/linux/man-pages/man5/resolv.conf.5.html)).

This issue can be reproduced by enabling the verbose output (which outputs the received UDP request) of the unit test `SearchDomainsServFailOnAAAA`.

*Before:*
```
$ ./arestest -v --gtest_filter=AddressFamiliesAI/MockChannelTestAI.SearchDomainsServFailOnAAAA/0
Note: Google Test filter = AddressFamiliesAI/MockChannelTestAI.SearchDomainsServFailOnAAAA/0
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from AddressFamiliesAI/MockChannelTestAI
[ RUN      ] AddressFamiliesAI/MockChannelTestAI.SearchDomainsServFailOnAAAA/0
Configured IPv4 mock server with TCP socket 3 on port 5300 and UDP socket 4 on port 5300
Configured library with servers: 127.0.0.1:5300
received UDP request REQ QRY RD  Q:{'www.first.com' IN A} on port 5300
ProcessRequest(27360, 'www.first.com', A)
sending reply RSP QRY AA NXDOMAIN Q:{'www.first.com' IN A} on port 5300
received UDP request REQ QRY RD  Q:{'www.first.com' IN AAAA} on port 5300
ProcessRequest(10278, 'www.first.com', AAAA)
sending reply RSP QRY AA NXDOMAIN Q:{'www.first.com' IN AAAA} on port 5300
received UDP request REQ QRY RD  Q:{'www.second.org' IN A} on port 5300
ProcessRequest(62567, 'www.second.org', A)
sending reply RSP QRY AA NOERROR Q:{'www.second.org' IN A} A:{'www.second.org' IN A TTL=512 2.3.4.5} on port 5300
received UDP request REQ QRY RD  Q:{'www.second.org' IN AAAA} on port 5300
ProcessRequest(30286, 'www.second.org', AAAA)
sending reply RSP QRY AA NXDOMAIN Q:{'www.second.org' IN AAAA} on port 5300
[       OK ] AddressFamiliesAI/MockChannelTestAI.SearchDomainsServFailOnAAAA/0 (2 ms)
```

*After:*
```
./arestest -v --gtest_filter=AddressFamiliesAI/MockChannelTestAI.SearchDomainsServFailOnAAAA/0
Note: Google Test filter = AddressFamiliesAI/MockChannelTestAI.SearchDomainsServFailOnAAAA/0
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from AddressFamiliesAI/MockChannelTestAI
[ RUN      ] AddressFamiliesAI/MockChannelTestAI.SearchDomainsServFailOnAAAA/0
Configured IPv4 mock server with TCP socket 3 on port 5300 and UDP socket 4 on port 5300
Configured library with servers: 127.0.0.1:5300
received UDP request REQ QRY RD  Q:{'www.first.com' IN AAAA} on port 5300
ProcessRequest(20810, 'www.first.com', AAAA)
sending reply RSP QRY AA NXDOMAIN Q:{'www.first.com' IN AAAA} on port 5300
received UDP request REQ QRY RD  Q:{'www.second.org' IN AAAA} on port 5300
ProcessRequest(16009, 'www.second.org', AAAA)
sending reply RSP QRY AA NXDOMAIN Q:{'www.second.org' IN AAAA} on port 5300
received UDP request REQ QRY RD  Q:{'www.third.gov' IN AAAA} on port 5300
ProcessRequest(38234, 'www.third.gov', AAAA)
sending reply RSP QRY AA SERVFAIL Q:{'www.third.gov' IN AAAA} on port 5300
received UDP request REQ QRY RD  Q:{'www.third.gov' IN AAAA} on port 5300
ProcessRequest(38234, 'www.third.gov', AAAA)
sending reply RSP QRY AA SERVFAIL Q:{'www.third.gov' IN AAAA} on port 5300
received UDP request REQ QRY RD  Q:{'www.third.gov' IN AAAA} on port 5300
ProcessRequest(38234, 'www.third.gov', AAAA)
sending reply RSP QRY AA SERVFAIL Q:{'www.third.gov' IN AAAA} on port 5300
received UDP request REQ QRY RD  Q:{'www.first.com' IN A} on port 5300
ProcessRequest(26330, 'www.first.com', A)
sending reply RSP QRY AA NXDOMAIN Q:{'www.first.com' IN A} on port 5300
received UDP request REQ QRY RD  Q:{'www.second.org' IN A} on port 5300
ProcessRequest(60572, 'www.second.org', A)
sending reply RSP QRY AA NOERROR Q:{'www.second.org' IN A} A:{'www.second.org' IN A TTL=512 2.3.4.5} on port 5300
AddrInfoCallback({ARES_SUCCESS {addr=[2.3.4.5]}})
[       OK ] AddressFamiliesAI/MockChannelTestAI.SearchDomainsServFailOnAAAA/0 (2 ms)
```

FYI: @bradh352 @ki11roy

**Comment by bradh352** [comment]
@ChristianAmmer yes, I'd definitely like to get this addressed before the next release of c-ares as these queries should definitely be in parallel.  I know you wrote the original code, but then another contributor came and made it more standards compliant but serial lookups.  If you can submit a PR to bring back parallelism, that'd be great.  I'm currently overloaded so won't have time for a while.

**Comment by ChristianAmmer** [comment]
Hi @bradh352, I will have a look how much effort it would be, I'm not familar with the current implementation anymore. At first look, I'm wondering why the [`ares_addrinfo`](https://github.com/ChristianAmmer/c-ares/blob/master/ares.h#L572-L581) definition, which was conforming to the [RFC definition](https://tools.ietf.org/html/rfc3493#page-23), was changed to a propriatary [three structures](https://github.com/c-ares/c-ares/blob/master/ares.h#L585-L611) variant. What was the rational therefore – #257 does not make it clear.

**Comment by ki11roy** [comment]
Original implementation of that code was sequential, but broken for getaddrinfo, so #257 fixed that among the other issues. Regretfully I had no time to properly thought on how to get the parallelization feature by @ChristianAmmer back, but I can join the discussion in a week or so as it is a blocker for release.
As for the rationale behind the structure please see discussion in #257

**Comment by bradh352** [comment]
The "standard" doesn't offer things like ability to access TTLs nor CNAMEs.  These were things people had requested, and it makes sense to be able to provide those via an API that is "similar" to getaddrinfo().  Of course there is no guarantee every system c-ares runs on has an addrinfo structure to begin with, so if we must define our own, then we can also add our own extensions.

**Comment by ChristianAmmer** [comment]
From RFC 1034
> CNAME identifies the canonical name of an alias

and the canonical name is part of the RFC addressinfo structure as member `ai_canonname`. Maybe I'm missing something but in my understanding it is part of the structure.
Sure TTL is missing but this could just be an additional "non standard" member extension.

**Comment by bradh352** [comment]
Sure, but you could have multiple cnames in a chain, but the rfc only gives you one of them.

**Comment by ki11roy** [comment]
Moreover, you can have TTL for each CNAME in a chain. TTL was added to substructures ares_addrinfo_node and ares_addrinfo_cname to address this issue. Can’t see the profits of being compliant with RFCs addrinfo for the library and agree with @bradh352 in that sense.
Anyway, how it is connected with the original issue of parallel queries?

**Comment by bradh352** [comment]
Fixed by #290
